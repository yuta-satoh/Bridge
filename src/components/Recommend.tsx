import Link from "next/link";
import Image from "next/image";
import useSWR, { Fetcher } from 'swr';

type Item = {
  id: number,
  name: string,
  description: string,
  genre: string,
  category: string,
  price: number,
  imgurl: string,
  stock: number,
  delete: boolean,
}

export default function Recommend({ filteredItemData }: { filteredItemData: Item[] }) {
  const fetcher: Fetcher<Item[], string> = (...args) => fetch(...args).then((res) => res.json());
  
  const endCartItem = filteredItemData[filteredItemData.length - 1]
  const itemQuery = filteredItemData.reduce((query, item) => query + `,id.eq.${item.id}`, "").replace(",", "")
  const { data: recommendItem, error } = useSWR(`/api/items?genre=eq.${endCartItem.genre}&not.or=(${itemQuery})`, fetcher)
  console.log(recommendItem)

  if (error) return (
    <div className="w-4/5 mx-auto">
      <h2 className="text-3xl font-bold mt-10">カート</h2>
      <div className="block py-40 mx-auto my-10 bg-neutral-100 rounded">
        <h2 className="text-center text-3xl font-bold mb-10">おすすめ商品がありません</h2>
      </div>
    </div>
  )
  if (!recommendItem) return <div>loading...</div>

  return (
    <div className="flex gap-3 mb-10">
      {recommendItem.slice(0, 2).map((item, index) => (
        <div key={index} className="border border-neutral-900 rounded p-3 mx-auto">
          <div className="flex gap-5">
            <Link href={'/'}>
              <Image src={item.imgurl} alt={item.name} width={120} height={120} className="rounded"/>
            </Link>
            <div className="px-2 py-2">
              <Link href={'/'}>
                <p className="underline mb-1">{item.name}</p>
                <p className="mt-1">¥ {item.price.toLocaleString()}</p>
              </Link>
              <button className="p-1.5 text-center text-white h-10 mt-6 rounded bg-blue-500 hover:bg-blue-700">
                カートに入れる
              </button>
            </div>
          </div>
        </div>  
      ))}
    </div> 
  )
}
