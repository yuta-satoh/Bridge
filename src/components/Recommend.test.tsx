import { render, screen } from "@testing-library/react";
import { Middleware, SWRResponse, SWRConfig } from "swr";
import userEvent from "@testing-library/user-event";

import Recommend from "./Recommend";

jest.mock('next/router', () => ({
    useRouter() {
      return {
        replace: jest.fn(),
        push: jest.fn(),
        reload: jest.fn(),
      };
    },
}));

const fetchMockOk = () => {
    return new Promise((resolve) => {
        resolve({
            ok: true,
            status: 200,
            statusText: 'OK',
        })
    })
}

describe('Recommendコンポーネント', () => {
    test('おすすめ商品のローディング中', () => {
        const cartMiddleware: Middleware = () => {
            return (): SWRResponse<any, any> => {
                return {
                    data: undefined,
                    error: undefined,
                    mutate: (_) => Promise.resolve(),
                    isValidating: false,
                    isLoading: false,
                }
            }
        }
        const { asFragment } = render(
            <SWRConfig value={{ use: [cartMiddleware] }}>
                <Recommend recommend={[{ id: 1, genre: 'ジャンル' }]} />
            </SWRConfig>
        );
        expect(asFragment()).toMatchSnapshot();
    })

    test('ユーザーカートのおすすめ商品が無い場合', () => {
        const cartMiddleware: Middleware = () => {
            return (): SWRResponse<any, any> => {
                return {
                    data: [],
                    error: Error(),
                    mutate: (_) => Promise.resolve(),
                    isValidating: false,
                    isLoading: false,
                }
            }
        }
        const { asFragment } = render(
            <SWRConfig value={{ use: [cartMiddleware] }}>
                <Recommend recommend={[{ id: 1, genre: 'ジャンル' }]} />
            </SWRConfig>
        );
        expect(asFragment()).toMatchSnapshot();
    })

    test('ユーザーカートのおすすめ商品が正常に描画される', () => {
        const cartMiddleware: Middleware = () => {
            return (): SWRResponse<any, any> => {
                return {
                    data: [
                        {
                            id: 1,
                            name: "ベルベットアームチェア",
                            description: "ベルベットのお洒落なデザインチェア。",
                            genre: "フェミニン",
                            category: "椅子",
                            price: 16000,
                            imgurl: "/images/chair/chair_feminine_1.jpg",
                            stock: 72,
                            delete: false
                        },
                        {
                            id: 2,
                            name: "丸椅子スツール",
                            description: "落ち着いた可愛らしいベロア風スツール。",
                            genre: "フェミニン",
                            category: "椅子",
                            price: 10000,
                            imgurl: "/images/chair/chair_feminine_2.jpg",
                            stock: 99,
                            delete: false
                        },
                    ],
                    error: undefined,
                    mutate: (_) => Promise.resolve(),
                    isValidating: false,
                    isLoading: false,
                }
            }
        }
        const { asFragment } = render(
            <SWRConfig value={{ use: [cartMiddleware] }}>
                <Recommend recommend={[{ id: 3, genre: 'フェミニン' }]} />
            </SWRConfig>
        );
        expect(asFragment()).toMatchSnapshot();
    })
    test('カートにおすすめ商品を追加する', async () => {
        global.fetch = jest.fn().mockImplementation(fetchMockOk);
        const user = userEvent.setup();

        const cartMiddleware: Middleware = () => {
            return (): SWRResponse<any, any> => {
                return {
                    data: [
                        {
                            id: 1,
                            name: "ベルベットアームチェア",
                            description: "ベルベットのお洒落なデザインチェア。",
                            genre: "フェミニン",
                            category: "椅子",
                            price: 16000,
                            imgurl: "/images/chair/chair_feminine_1.jpg",
                            stock: 72,
                            delete: false
                        },
                    ],
                    error: undefined,
                    mutate: (_) => Promise.resolve(),
                    isValidating: false,
                    isLoading: false,
                }
            }
        }
        const { asFragment } = render(
            <SWRConfig value={{ use: [cartMiddleware] }}>
                <Recommend recommend={[{ id: 3, genre: 'フェミニン' }]} />
            </SWRConfig>
        );
        const addBtn = screen.getByRole('button');
        // クリックさせてみたけどその後の処理を確認する方法がわからないのでここでストップ
        await user.click(addBtn);
        expect(asFragment()).toMatchSnapshot();
    })

})
