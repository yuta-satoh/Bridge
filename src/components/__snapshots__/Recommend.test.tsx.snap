// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Recommendコンポーネント おすすめ商品のローディング中 1`] = `
<DocumentFragment>
  <div />
</DocumentFragment>
`;

exports[`Recommendコンポーネント カートにおすすめ商品を追加する 1`] = `
<DocumentFragment>
  <div
    class="mt-16 mb-5"
  >
    <h2
      class="text-3xl font-bold"
    >
      こちらもいかがですか？
    </h2>
  </div>
  <div
    class="flex gap-3 mb-10"
  >
    <div
      class="border border-neutral-900 w-1/2 rounded p-3 mx-auto"
    >
      <div
        class="flex gap-5"
      >
        <a
          href="/items/itemlist/1"
        >
          <img
            alt="ベルベットアームチェア"
            class="rounded"
            data-nimg="1"
            decoding="async"
            height="100"
            loading="lazy"
            src="/_next/image?url=%2Fimages%2Fchair%2Fchair_feminine_1.jpg&w=256&q=75"
            srcset="/_next/image?url=%2Fimages%2Fchair%2Fchair_feminine_1.jpg&w=128&q=75 1x, /_next/image?url=%2Fimages%2Fchair%2Fchair_feminine_1.jpg&w=256&q=75 2x"
            style="color: transparent;"
            width="100"
          />
        </a>
        <div
          class="px-2 py-2"
        >
          <a
            href="/items/itemlist/1"
          >
            <p
              class="underline mb-1 text-sm"
            >
              ベルベットアームチェア
            </p>
            <p
              class="mt-1 mb-5"
            >
              ¥ 16,000
            </p>
          </a>
          <button
            class="jsx-896052889"
            type="button"
          >
            カートに入れる
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Recommendコンポーネント ユーザーカートのおすすめ商品が正常に描画される 1`] = `
<DocumentFragment>
  <div
    class="mt-16 mb-5"
  >
    <h2
      class="text-3xl font-bold"
    >
      こちらもいかがですか？
    </h2>
  </div>
  <div
    class="flex gap-3 mb-10"
  >
    <div
      class="border border-neutral-900 w-1/2 rounded p-3 mx-auto"
    >
      <div
        class="flex gap-5"
      >
        <a
          href="/items/itemlist/1"
        >
          <img
            alt="ベルベットアームチェア"
            class="rounded"
            data-nimg="1"
            decoding="async"
            height="100"
            loading="lazy"
            src="/_next/image?url=%2Fimages%2Fchair%2Fchair_feminine_1.jpg&w=256&q=75"
            srcset="/_next/image?url=%2Fimages%2Fchair%2Fchair_feminine_1.jpg&w=128&q=75 1x, /_next/image?url=%2Fimages%2Fchair%2Fchair_feminine_1.jpg&w=256&q=75 2x"
            style="color: transparent;"
            width="100"
          />
        </a>
        <div
          class="px-2 py-2"
        >
          <a
            href="/items/itemlist/1"
          >
            <p
              class="underline mb-1 text-sm"
            >
              ベルベットアームチェア
            </p>
            <p
              class="mt-1 mb-5"
            >
              ¥ 16,000
            </p>
          </a>
          <button
            class="jsx-896052889"
            type="button"
          >
            カートに入れる
          </button>
        </div>
      </div>
    </div>
    <div
      class="border border-neutral-900 w-1/2 rounded p-3 mx-auto"
    >
      <div
        class="flex gap-5"
      >
        <a
          href="/items/itemlist/2"
        >
          <img
            alt="丸椅子スツール"
            class="rounded"
            data-nimg="1"
            decoding="async"
            height="100"
            loading="lazy"
            src="/_next/image?url=%2Fimages%2Fchair%2Fchair_feminine_2.jpg&w=256&q=75"
            srcset="/_next/image?url=%2Fimages%2Fchair%2Fchair_feminine_2.jpg&w=128&q=75 1x, /_next/image?url=%2Fimages%2Fchair%2Fchair_feminine_2.jpg&w=256&q=75 2x"
            style="color: transparent;"
            width="100"
          />
        </a>
        <div
          class="px-2 py-2"
        >
          <a
            href="/items/itemlist/2"
          >
            <p
              class="underline mb-1 text-sm"
            >
              丸椅子スツール
            </p>
            <p
              class="mt-1 mb-5"
            >
              ¥ 10,000
            </p>
          </a>
          <button
            class="jsx-896052889"
            type="button"
          >
            カートに入れる
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Recommendコンポーネント ユーザーカートのおすすめ商品が無い場合 1`] = `
<DocumentFragment>
  <div
    class="w-4/5 mx-auto"
  >
    <div
      class="block py-40 mx-auto my-10 bg-neutral-100 rounded"
    >
      <h2
        class="text-center text-3xl font-bold mb-10"
      >
        おすすめ商品がありません
      </h2>
    </div>
  </div>
</DocumentFragment>
`;
